lib/SILOptimizer/Utils/ConstExpr.cpp

```diff
@@ -134,10 +132,42 @@ Type ConstExprFunctionState::simplifyType(Type ty) {
   return substitutionMap.empty() ? ty : ty.subst(substitutionMap);
 }
 
+// TODO: refactor this out somewhere sharable between autodiff and this code.
+static void lookupOrLinkWitnessTable(ProtocolConformanceRef confRef,
```

Comment:
I guess I don't really understand why this PR is adding so much new code (lookupOrLinkWitnessTable() and getWitnessMethodSubstitutions()) when very similar logic already exists in the devirtualizer.

You bring up a very good point that while the devirtualizer rewrites instructions in place, the evaluator looks at but does not mutate the AST. However it should be possible to come up with some common pieces that can be reused -- for example, a common utility returning the SILFunction and SubstitutionMap for a devirtualized call. The optimizer can build an ApplyInst from that, and the evaluator can proceed to evaluate the body of the SILFunction.

I'm generally pessimistic about PRs that add TODO/FIXME comments because in practice we rarely see them get addressed, as everyone is keen to move on to the next feature...

Reply:
Sounds good!

Reply:
I have some suggestions for addressing @slavapestov 's concerns.   It seems to me that `lookupOrLinkWitnessTable()` functionality is partly available in the `SILLinkerVisitor::visitProtocolConformance`. Apparently, this is the only other place where we _create_ a witness table declaration for a protocol conformance in case we do not find a witness table for the conformance. (Even though somewhat similar logic exist in other places, it seems to me that there is an assumption that the witness table declaration must exist for a protocol in other places.)  But, much of what visitProtocolConformance does, such as deserializing any witness methods and  visiting related protocol conformances, are not needed for the evaluator. (Deserialization of the witness method would happen later during evaluation, only if the function is actually invoked during evaluation.)  Therefore, it seems to me that  we could factor out this common functionality, shown below, from `visitProtocolConformance`.

```swift
SILWitnessTable *lookupOrCreateWitnessTable(ProtocolConformance *C, bool mustDeserialize,           
    SILModule &Mod) {
  SILWitnessTable *WT = Mod.lookUpWitnessTable(C, true);

  // If we don't find any witness table for the conformance,
  // try to create a witness table declaration and try again.
  if (!WT) {
    Mod.createWitnessTableDeclaration(C,
        getLinkageForProtocolConformance(
                          C->getRootNormalConformance(), NotForDefinition));

    // Adding the declaration may allow us to now deserialize the body.
    // Force the body if we must deserialize this witness table.
    if (mustDeserialize) {
      WT = Mod.lookUpWitnessTable(C, true);
      assert(WT && WT->isDefinition()
             && "unable to deserialize witness table when we must?!");
    } else {
      return WT;
    }
  }
  return WT;
}
```
I am not sure where this function should go (Perhaps in the SILLinker.cpp itself or SILModule.cpp?).  If we have this factored out, both the evaluator and `SILLinkerVisitor::visitProtocolConformance` can use it.

Reply:
Check out https://github.com/apple/swift/pull/21476 -- I updated lookUpWitnessTable() to create the declaration if one doesn't exist, so you can simplify lookupOrLinkWitnessTable() or perhaps remove it entirely.

Reply:
Thanks a lot @slavapestov. This is awesome.